src/realm/views.cpp

```diff
@@ -215,38 +336,68 @@ void RowIndexes::do_sort(const SortDescriptor& order, const SortDescriptor& dist
             ++detached_ref_count;
     }
 
-    if (distinct) {
-        auto sorting_predicate = distinct.sorter(m_row_indexes);
-
-        // Remove all rows which have a null link along the way to the distinct columns
-        if (sorting_predicate.has_links()) {
-            v.erase(std::remove_if(v.begin(), v.end(),
-                                   [&](auto&& index) { return sorting_predicate.any_is_null(index); }),
-                    v.end());
+    const size_t num_descriptors = ordering.size();
+    for (size_t desc_ndx = 0; desc_ndx < num_descriptors; ++desc_ndx) {
+        if (ordering.descriptor_is_sort(desc_ndx)) {
+            // an empty (placeholder) sort descriptor is a no-op
+            if (!ordering[desc_ndx]) continue;
+
+            SortDescriptor::Sorter sort_predicate = ordering[desc_ndx].sorter(m_row_indexes);
+
+            std::sort(v.begin(), v.end(), std::ref(sort_predicate));
+
+            bool is_last_ordering = desc_ndx >= num_descriptors - 1;
+            // not doing this on the last step is an optimisation
+            if (!is_last_ordering) {
+                const size_t v_size = v.size();
+                // Distinct must choose the winning unique elements by sorted
+                // order not by the previous tableview order, the lowest
+                // "index_in_view" wins.
+                for (size_t i = 0; i < v_size; ++i) {
+                    v[i].index_in_view = i;
+                }
+            }
         }
+        else { // distinct descriptor
+            REALM_ASSERT_DEBUG(ordering.descriptor_is_distinct(desc_ndx));
 
-        // Sort by the columns to distinct on
-        std::sort(v.begin(), v.end(), std::ref(sorting_predicate));
+            // an empty distinct descriptor is a no-op
+            if (!ordering[desc_ndx]) continue;
 
-        // Remove all duplicates
-        v.erase(std::unique(v.begin(), v.end(),
-                            [&](auto&& a, auto&& b) {
-                                // "not less than" is "equal" since they're sorted
-                                return !sorting_predicate(a, b, false);
-                            }),
-                v.end());
+            // Setting an order on the distinct descriptor is now incorrect.
+            // Distinct uses the order of the view. If the order matters,
+            // then do a sort beforehand to change the results of distinct.
+            if (ordering[desc_ndx].has_custom_order()) {
```

Comment:
Because at this point, we aren't sure if the descriptor is used for sorting (which allows this) or for distinct (which doesn't). Hmmm - also solvable by creating different classes for each.